app/frontend/src/features/search/users.ts

```diff
@@ -128,24 +121,38 @@ export const addClusteredUsersToMap = (
 ) => {
   map.addSource("clustered-users", sources["clustered-users"]);
   addPinImages(map);
-  map.addLayer(layers["clusterLayer"]);
-  map.addLayer(layers["clusterCountLayer"]);
-  map.addLayer(layers["unclusteredPointLayer"]);
+  map.addLayer(layers.clusterLayer);
+  map.addLayer(layers.clusterCountLayer);
+  map.addLayer(layers.unclusteredPointLayer);
   if (userClickedCallback) {
-    map.on("click", layers["unclusteredPointLayer"].id, userClickedCallback);
+    map.on("click", layers.unclusteredPointLayer.id, userClickedCallback);
   }
-  map.on("click", layers["clusterLayer"].id, zoomCluster);
+  map.on("click", layers.clusterLayer.id, zoomCluster);
 };
 
-export const addUsersToMap = (
+export const filterUsers = (
   map: MaplibreMap,
+  ids: number[] | null,
   userClickedCallback?: MapClickedCallback
 ) => {
-  map.addSource("all-objects", sources["all-objects"]);
-  addPinImages(map);
-  map.addLayer(layers["users"]);
-
+  //clusters can only be filtered at the source before rendering
+  //so we have to remove the layers and sources and re-add
   if (userClickedCallback) {
-    map.on("click", layers["users"].id, userClickedCallback);
+    map.off("click", layers.unclusteredPointLayer.id, userClickedCallback);
+    map.off("click", layers.clusterLayer.id, zoomCluster);
   }
+  map.removeLayer(layers.clusterLayer.id);
+  map.removeLayer(layers.clusterCountLayer.id);
+  map.removeLayer(layers.unclusteredPointLayer.id);
+  map.removeSource("clustered-users");
+
+  if (ids) {
+    //@ts-ignore - type definition incorrect
+    sources["clustered-users"].filter = ["in", ["get", "id"], ["literal", ids]];
```

Comment:
All these Mapbox GL specific config stuff reads like black magic to me 🙈

Reply:
I can put comments in. This is just `ids contains(sourceItem.id)`